### ISSUE TITLE
про коллбэки

### ISSUE THREAD

**Issue by pansa** [tmp1test/t2#1]
Привет!

"Предусмотреть возможность вызова callback'ов для pre- и -post-обрабоки загружаемых файлов. - Не до конца понял, что такое пре/пост обработка загружаемых файлов"

Например, мы хотим передать загруженный файл на проверку антивирусному сканеру. Соответственно, нужен вызов, в котором мы могли бы реализовать какие-либо действия над только что загруженным файлом. Это -post.

Вариант с -pre-коллбэком. Например, хотим выполнить авторизацию - а можно ли этому юзеру загружать файлы. Или  проверить лимит по размеру. Т.е это момент, когда файл принят/загружается, но еще не сохранен в основное хранилище.

Если есть вопросы - пишите. =)

**Comment by pansa** [comment]
Да, пример с авторизацией - отлично. Это действително синхронная операция. Если есть желание, будет интересна ваша реализация. При этом сам бэкэнд авторизации не нужно, можно просто оставить заглушку, внутри которой с некоторой задержкой происходит авторизация.

Асинхронный пост-аплоад тоже будет интересен. Ответ клиенту в этом случае возвращается сразу по факту успешной заливки файла, вызов коллбэка - асинхронный.

**Comment by tmp1test** [comment]
@pansa 
добавил примеры коллбеков

у структуры handler можно посмотреть метод `checkAuthorization`
Метод проверяет авторизацию просто сравнением токена из заголовка со строкой и эмулирует сетевую задержку в 200мс

Метод checkFile добавляет файл на проверку (сортированное множество в редисе). Потом оттуда можно вычитывать задачи и исполнять их. В зависимости от результата - принимать или не принимать какие-либо действия с файлом.

Правильно ли я понял вас?

**Comment by pansa** [comment]
> добавил примеры коллбеков
Это не совсем коллбеки =(

Смотрите, коллбеками мы хотим дать возможность другим разработчикам, которые будут использовать этого демона, добавлять новую функциональность , не трогая основной код.

**Comment by tmp1test** [comment]
@pansa 
Добавил коллбеки в конструктор Application

Добавил тесты на коллбеки в handler_test.go - TestHandlerUploadFileCallbacks
